### DISCUSSION TITLE
[Board]: Affordable, portable 1W Meshtastic ready platform with ESP32+OLED+GPS pads (Radiomaster Bandit Nano 900MHz ELRS Module)

### DISCUSSION THREAD

**Discussion by uberhalit** [meshtastic/firmware#3794] [Ideas]
### SOC

ESP32

### Lora IC

SX1276

### Product Link

https://www.radiomasterrc.com/products/bandit-nano-expresslrs-rf-module

### Description

Being a drone racer, I've always wondered if we could turn some ELRS 900MHz gear into a low-cost, reliable, and super powerful Meshtastic device.

By adhering to the [ELRS board definitions](https://github.com/ExpressLRS/targets/blob/master/TX/Radiomaster%20Bandit%20Micro.json), I was able to incorporate base support for the model referred to above:
![ser](https://github.com/meshtastic/firmware/assets/19159295/d23cc639-ba7f-4589-8d2a-6f034e70d5f3)
It even features free UART pads that can be utilized to potentially connect a GPS.
[Fork available here.](https://github.com/uberhalit/firmware/commit/8397551f24b6ee83d6f61de684dcbd1d6046508c)
The display requires a flipping in settings.

Nevertheless, currently, neither of the two PAs appears to be engaged, and the power output is below -20dBm. I attempted to hack together a basic 30dbm output PoC, but was unable to do so. 

[ELRS is calling `dacWrite()` on pin 26 (power_apc2)](https://github.com/ExpressLRS/ExpressLRS/blob/66998e173252c4c0a9e6e0e89010761f56ec9ff2/src/lib/POWERMGNT/POWERMGNT.cpp#L298) and a regular [`writeRegister` to register 0x09 to activate the PAs.](https://github.com/ExpressLRS/ExpressLRS/blob/66998e173252c4c0a9e6e0e89010761f56ec9ff2/src/lib/SX127xDriver/SX127x.cpp#L263)
The [call to reg 0x09 is already being executed](https://github.com/jgromes/RadioLib/blob/9c6eb20169b1019ca4215f6756a7fd32be2649b0/src/modules/SX127x/SX1278.cpp#L336) in the current firmware implementation, albeit with different settings.
Manually setting both to the same values as seen in ELRS source does not seem to work, I'd assume that either the SX1276 support is incomplete or a crucial PA initialization is missing (mostly related to [RFAMP](https://github.com/ExpressLRS/ExpressLRS/blob/3.3.2/src/lib/RFAMP/RFAMP_hal.cpp) or [LR1121](https://github.com/ExpressLRS/ExpressLRS/blob/66998e173252c4c0a9e6e0e89010761f56ec9ff2/src/lib/LR1121Driver/LR1121.cpp#L296)). 

In any case, Radiomaster is known for their excellent support. They immediately responded and offered technical help, so this board should come to live eventually.

**Comment by RadioMasterRC** [comment]
Thanks for your efforts @uberhalit !

If anyone is interested in contributing, we are happy to make hardware available. Reach out to @sachaw on Discord and we can hook you up.

Looking forward to it 💯 

Ben.

**Comment by dhandzic** [comment]
Hmm, the SX1276 supports +20dBm and the LR1121 +22 dBm output. Unless there is an external PA it is hard to see how +30dBm could be achieved. Any information if there is an external PA and how it is controlled or if it is just a fixed gain PA?

**Comment by RadioMasterRC** [comment]
> Hmm, the SX1276 supports +20dBm and the LR1121 +22 dBm output. Unless there is an external PA it is hard to see how +30dBm could be achieved. Any information if there is an external PA and how it is controlled or if it is just a fixed gain PA?

Hi @dhandzic The Bandit Nano uses a Skyworks SKY66122 PA. The PA Voltage is controlled via the ESP32 DAC port and works together with the SX1276 RF power to set the output power. The effective power control range is 100mw - 1000mw

Additional info:
https://github.com/ExpressLRS/ExpressLRS/tree/master/src/lib/POWERMGNT

The Bandit Nano uses the "DACWRITE" function.

**Comment by dhandzic** [comment]
> > Hmm, the SX1276 supports +20dBm and the LR1121 +22 dBm output. Unless there is an external PA it is hard to see how +30dBm could be achieved. Any information if there is an external PA and how it is controlled or if it is just a fixed gain PA?
> 
> Hi @dhandzic The Bandit Nano uses a Skyworks SKY66122 PA. The PA Voltage is controlled via the ESP32 DAC port and works together with the SX1276 RF power to set the output power. The effective power control range is 100mw - 1000mw
> 
> Additional info: https://github.com/ExpressLRS/ExpressLRS/tree/master/src/lib/POWERMGNT
> 
> The Bandit Nano uses the "DACWRITE" function.

Thank you for the information. Cool with a +30 dBm node as a repeater. Do you have any figures on the standby current consumption?

**Comment by dhandzic** [comment]
> > Hmm, the SX1276 supports +20dBm and the LR1121 +22 dBm output. Unless there is an external PA it is hard to see how +30dBm could be achieved. Any information if there is an external PA and how it is controlled or if it is just a fixed gain PA?
> 
> Hi @dhandzic The Bandit Nano uses a Skyworks SKY66122 PA. The PA Voltage is controlled via the ESP32 DAC port and works together with the SX1276 RF power to set the output power. The effective power control range is 100mw - 1000mw
> 
> Additional info: https://github.com/ExpressLRS/ExpressLRS/tree/master/src/lib/POWERMGNT
> 
> The Bandit Nano uses the "DACWRITE" function.

Could the device be used for 869 MHz too or is the RX filter tuned for 915 MHz?

**Comment by uberhalit** [comment]
> > > Hmm, the SX1276 supports +20dBm and the LR1121 +22 dBm output. Unless there is an external PA it is hard to see how +30dBm could be achieved. Any information if there is an external PA and how it is controlled or if it is just a fixed gain PA?
> > 
> > 
> > Hi @dhandzic The Bandit Nano uses a Skyworks SKY66122 PA. The PA Voltage is controlled via the ESP32 DAC port and works together with the SX1276 RF power to set the output power. The effective power control range is 100mw - 1000mw
> > Additional info: https://github.com/ExpressLRS/ExpressLRS/tree/master/src/lib/POWERMGNT
> > The Bandit Nano uses the "DACWRITE" function.
> 
> Thank you for the information. Cool with a +30 dBm node as a repeater. Do you have any figures on the standby current consumption?

I measured 2.3W on ELRS while transmitting 30.1dbm and around 0.5W in idle with display on. 

> Could the device be used for 869 MHz too or is the RX filter tuned for 915 MHz?

You can set it to 868MHz too, that's what I am using it with.

**Comment by thebentern** [comment]
If somebody wants to add official support for this, we can for sure integrate it, but I'm closing this as an active issue. I don't have the bandwidth to integrate more SX1276 based boards with all of the backlog of devices to support.

**Comment by uberhalit** [comment]
> If somebody wants to add official support for this, we can for sure integrate it, but I'm closing this as an active issue. I don't have the bandwidth to integrate more SX1276 based boards with all of the backlog of devices to support.

Is this subject open to discussion? As far as I am aware, this device should have high priority, as it is the only portable 900MHz device with adequate output power for dense urban environments. I've noticed a lot of interest lately, and the top complaint in YT comments is the poor range in cities without them realizing that their device is limited to 125mW (I guess the 30dbm max output setting is a bit misleading). It would be unfortunate if potential users were to abandon the project because they were unable to obtain a proper link in complex environments.

**Comment by gjelsoe** [comment]
> > If somebody wants to add official support for this, we can for sure integrate it, but I'm closing this as an active issue. I don't have the bandwidth to integrate more SX1276 based boards with all of the backlog of devices to support.
> 
> Is this subject open to discussion? As far as I am aware, this device should have high priority, as it is the only portable 900MHz device with adequate output power for dense urban environments. I've noticed a lot of interest lately, and the top complaint in YT comments is the poor range in cities without them realizing that their device is limited to 125mW (I guess the 30dbm max output setting is a bit misleading). It would be unfortunate if potential users were to abandon the project because they were unable to obtain a proper link in complex environments.

Are you still working on this?

**Comment by thebentern** [comment]
> it is the only portable 900MHz device with adequate output power for dense urban environments

The G2 station can achieve more power than this radio, and has an SX1262 which features a much more sensitive RX end, which makes a greater difference than raw transmit power in my experience, with multiple A/B tests of SX1262 with RX boosted gain w/ 22dbm TX vs 30dbm setups. I'm afraid 30dbm will not be the silver bullet expectations would have you believe. This is a line of sight technology.

**Comment by KingRikkie** [comment]
> > it is the only portable 900MHz device with adequate output power for dense urban environments
> 
> The G2 station can achieve more power than this radio, and has an SX1262 which features a much more sensitive RX end, which makes a greater difference than raw transmit power in my experience, with multiple A/B tests of SX1262 with RX boosted gain w/ 22dbm TX vs 30dbm setups. I'm afraid 30dbm will not be the silver bullet expectations would have you believe. This is a line of sight technology.

Sorry for squeezing myself into this convo but the G2 "stationary" is not really a comparable portable device. It's 3 times the size, more than double the price (before taxes) and currently unavailable. I much prefer to use existing ELRS hardware that serves multiple purposes.

**Comment by thebentern** [comment]
More expensive, I'll grant you. But it is a quite portable device.  
![image](https://github.com/meshtastic/firmware/assets/9000580/a77ec4b2-6d0c-4d9e-bc7d-ec9198a54888)

**Comment by KingRikkie** [comment]
> More expensive, I'll grant you. But it is a quite portable device. 

Although still portable, this kind of confirms how much bigger it is compared to the Bandit. I am unaware of your involvement in the G2 project or its vendor, however, the cost is beyond reasonable to me. Including import fees and shipping, this comes up to 150€ + another 10€ for an 868MHz antenna. My local FPV shop has the Bandit on sale for 42€, so I could buy four Bandits for the price of one G2. I'm sorry, but this isn't in any way an alternative.

**Comment by uberhalit** [comment]
Regardless of dimensions and price, I think we can all agree that there is a profound lack of higher power devices. Moreover, it is unquestionable that the majority of users would prefer to utilize a device they might already own rather than buying a costly new one. 

> Are you still working on this?

Unfortunately my university schedule is currently too busy to continue. Feel free to pick it up.

**Comment by gjelsoe** [comment]
> > Are you still working on this?
> 
> Unfortunately my university schedule is currently too busy to continue. Feel free to pick it up.

Tried too, as [RadioMasterRC](https://github.com/RadioMasterRC) wrote but nothing happened.

**Comment by spcano01** [comment]
Larger, unavailable, 4X more expensive, and not available mainstream (single consulting group in China - nice guys though!) 

It would be great to have this as an option, as I can easily see this as the easiest/cheapest/most-consumer-friendly (like T-Echo) device available.

**Comment by garthvh** [comment]
> Larger, unavailable, 4X more expensive, and not available mainstream (single consulting group in China - nice guys though!)
> 
> It would be great to have this as an option, as I can easily see this as the easiest/cheapest/most-consumer-friendly (like T-Echo) device available.

Nothing is stopping you from giving it a try, a 1276 with a fan is not a target any core contributors are very interested in.

**Comment by RadioMasterRC** [comment]
If anyone is interested in picking this up and needs hardware feel free to reach out, share your discord name and I can get in touch.

Cheers,

Ben.

**Reply by gjelsoe** [reply]
Discord : cossie1413

**Reply by garthvh** [reply]
> If anyone is interested in picking this up and needs hardware feel free to reach out, share your discord name and I can get in touch.
> 
> 
> 
> Cheers,
> 
> 
> 
> Ben.

Any chance of a 1262 model? We have found its rx boosted gain is much better than a 1276 for message delivery in meshtastic

**Comment by gjelsoe** [comment]
@uberhalit If you haven't notice it yet. As of version [2.3.11.2740a56 Alpha](https://github.com/meshtastic/firmware/releases/tag/v2.3.11.2740a56) there are support for the Bandit Nano.
~~You need to build it yourself for now, WEB Flasher is not updated yet.~~ Is added now.

I've found the it uses RFO, that's why you weren't able to get the PA working.

Stuff that still needs to be sorted.
5-Way button, works as one now.
FAN is on all the time, ability to turn it off would be nice.
Variable power output (10-30dbm), is at 30dbm now. (WIP, have some code but needs testing)

Michael.

**Reply by Uravass** [reply]
Is only the Bandit Nano supported, or will the "Micro" version work, as well?

**Comment by gjelsoe** [comment]
Micro is also supported.

Is only the Bandit Nano supported, or will the "Micro" version work, as

**Reply by Uravass** [reply]
Thank you - I'd imagine this means that the big metal one works, as well?  I know I should've asked this before....

**Reply by gjelsoe** [reply]
Currently no, in the future maybe :)

The "Big Brother" doesn't bring much that Micro and Nano doesn't do and it's $50 more than the others.

**Reply by Uravass** [reply]
Thanks for your reply and contributions to the community - we appreciate it!  
I have the nano and a couple of micros (I fly FPV, too) and they're great, but I was just wondering about the big one since it seems to have better heat dissipation due to it's larger fan and aluminum case and may last longer when running at 1W around the clock.
Cheers -

**Reply by gjelsoe** [reply]
Good question, only way to figure that out is by testing.
But my guess is, they will do just fine. 

If needed I will add the support...

**Reply by Uravass** [reply]
My neighbor has his bandit nano up and running on our mesh - I'll flash one of my micros, here, momentarily and run all the same range tests that I did with my t-beam and report my findings... or any problems.

**Reply by jarno83** [reply]
> My neighbor has his bandit nano up and running on our mesh - I'll flash one of my micros, here, momentarily and run all the same range tests that I did with my t-beam and report my findings... or any problems.

Hi, how did it go? Thanks

**Reply by Uravass** [reply]
No problems - the nano is better suited to the application.  The duty cycle is so low that the unit barely gets warm at full output power - this was my biggest fear which has been totally ameliorated.

**Reply by Uravass** [reply]
I'll add the the unit's range & penetration are drastically improved vs the 100mW devices.

**Comment by dystic-bread** [comment]
It would be interesting to see support for the Nomad too, although the complexity of dual TX/RX on on ESP32, not sure what that looks like

**Reply by RobertoD91** [reply]
I have a Radiomaster Nomad for testing.
I made a quick build using the pin from the expresslrs repo. Someone has any advice before flashing?

**Reply by Uravass** [reply]
I don't see the benefit unless the lr1121 is better-performing in some way vs. the SX-1276.  It would be beneficial if an mLRS module with 1W output could be flashed - Matek has a version.  That has the SX-1262 chip (the same one most Meshtastic modules use) which IS better-performing by comparison to the SX-1276 on the Bandit.

**Reply by gjelsoe** [reply]
Have no experience with the Nomad, so I can't help here.

**Reply by RobertoD91** [reply]
I flashed a first, very experimental, version on the RadioMaster Nomad and the first good news is: it did not explode :)

I may open a dedicated discussion later, but here is a very early status update.
This is my first time playing with Meshtastic, so any feedback, even basic, is very welcome.

Current status:

- pins from the ExpressLRS repo: SPI, fan, buttons, led, etc.
- the second LR1121 is disabled
- 868 MHz was tested with a Heltec V4 and seems to work
- I tried testing 2.4 GHz with another ELRS transmitter (BetaFPV nano), but had no luck yet.
- since I do not fully understand the external PA yet, I kept it at very low power to avoid damage.
- adding only the `variants` folder was not enough. I found (and Claude fixed) a few issues

**Reply by RobertoD91** [reply]
> I don't see the benefit unless the lr1121 is better-performing in some way vs. the SX-1276. It would be beneficial if an mLRS module with 1W output could be flashed - Matek has a version. That has the SX-1262 chip (the same one most Meshtastic modules use) which IS better-performing by comparison to the SX-1276 on the Bandit.

I'm new to the LoRa/Meshtastic world, so I don't really know the performance differences between LR1121 and SX1276/SX1262/etc.

I already had a Nomad and was just curious to learn something new by trying to put Meshtastic on it. I'll mainly use the Nomad for drones/planes anyway :)

I don't have any matek elrs/mlrs tx :(

**Comment by johnmorningview** [comment]
On full sized module how can i get LED's working.  Everything on bandit nano module works great but on full sized module leds do nothing and both extra buttons do nothing (by default) (i'm not sure if they can be setup in config)
I flashed full module with web flasher with same firmware as bandit nano module. Had to change display settings. Can we get a firmware for the full sized version? it has cool led's and two buttons.

**Reply by Uravass** [reply]
Don't hold yer breath.  The full-size doesn't offer anything due to the already low duty-cycle of meshtastic... The nano never even gets warm... yeah, no leds, though...
